x/meicdp/handler.go

```diff
@@ -19,23 +23,122 @@ func NewHandler(keeper Keeper) sdk.Handler {
 		case channeltypes.MsgPacket:
 			var responseData oracle.OracleResponsePacketData
 			if err := types.ModuleCdc.UnmarshalJSON(msg.GetData(), &responseData); err == nil {
-				fmt.Println("I GOT DATA", responseData.Result, responseData.ResolveTime)
-				// handleOraclePacket(ctx, keeper, responseData)
+				err := handleOracleRespondPacketData(ctx, keeper, responseData)
+				if err != nil {
+					return nil, sdkerrors.Wrapf(
+						types.ErrRequestOracleData,
+						"error while handle response oracle data: %v",
+						err,
+					)
+				}
+
 				return &sdk.Result{Events: ctx.EventManager().Events().ToABCIEvents()}, nil
 			}
-			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal oracle packet data")
+			return nil, sdkerrors.Wrapf(
+				sdkerrors.ErrUnknownRequest,
+				"cannot unmarshal oracle packet data",
+			)
 
 		case MsgLockCollateral:
 			return handleMsgLockCollateral(ctx, keeper, msg)
 
+		case MsgUnlockCollateral:
+			msgCount := keeper.GetMsgCount(ctx)
+
+			// Setup oracle request
+			bandChainID := BandChainID
+			port := MeiCdpPort
+			oracleScriptID := oracle.OracleScriptID(OracleScriptID)
+			clientID := fmt.Sprintf("Msg:%d", msgCount)
+			calldata := encodeRequestParams(AtomSymbol, AtomMultiplier)
+			askCount := int64(1)
+			minCount := int64(1)
+
+			channelID, err := keeper.GetChannel(ctx, bandChainID, port)
+
+			dataRequest := types.NewDataRequest(
+				oracleScriptID,
+				channelID,
+				bandChainID,
+				port,
+				clientID,
+				calldata,
+				askCount,
+				minCount,
+				msg.Sender,
+			)
+
+			// Set message to the store for waiting the oracle response packet.
+			keeper.SetMsg(ctx, msgCount, msg)
+
+			err = requestOracle(ctx, keeper, dataRequest)
+			if err != nil {
+				return nil, sdkerrors.Wrapf(
+					types.ErrRequestOracleData,
+					"error while request oracle data: %v",
+					err,
+				)
+			}
+
+			return &sdk.Result{Events: ctx.EventManager().Events().ToABCIEvents()}, nil
+
+		case MsgBorrowDebt:
+			msgCount := keeper.GetMsgCount(ctx)
```

Comment:
This is duplicate of the above. Can't we handle them together?

Reply:
Added new handler function to handle together